### PR TITLE
[SPARK-5762] Fix shuffle write time for sort-based shuffle

### PR DESCRIPTION
**Pull request by kayousterhout**
@mateiz was excluding the time to write this final file from the shuffle write time intentional?

### CODE REVIEW

**Comment by sryza** [comment]
The shuffle write time is meant to be measured in nanoseconds, right?

Also, we should be measuring write time as well for the more common case where we're merging partition files.  Unless that's somehow measured already?

**Comment by kayousterhout** [comment]
Ah thanks for noticing that -- fixed!  And yeah the other case works fine because it goes through the DiskBlockObjectWriter (sounds like you figured this out).

**Comment by sryza** [comment]
My other concern is that we're also measuring the time spent reading the spill files from disk.  Though I suppose there's no way around this because we use `transferTo`?  If others are OK with that, then LGTM.

**Comment by mateiz** [comment]
I think this was just an oversight, it's good to include that time. The change looks good to me.

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 1]
```diff
@@ -740,6 +741,8 @@ private[spark] class ExternalSorter[K, V, C](
           in.close()
         }
       }
+      context.taskMetrics.shuffleWriteMetrics.map(
```
Comment (ksakellis) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: Should we write this metric if there was an exception thrown above? if so, maybe we should add this in the finally block. If not, then this is ok.

**Comment by ksakellis** [comment]
LGTM, @sryza I think its okay to include the reading time since this whole operation you can argue is for writing out the file - even if there is some reading.

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala (hunk 1 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: Good point -- done.

**Comment by kayousterhout** [comment]
I had the same thought as @ksakellis re:reading time.  Thanks for looking at this all!

**Comment by kayousterhout** [comment]
Jenkins, retest this please

File: core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala [hunk 2]
```diff
@@ -739,6 +740,8 @@ private[spark] class ExternalSorter[K, V, C](
         if (in != null) {
           in.close()
         }
+        context.taskMetrics.shuffleWriteMetrics.map(
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala]: could be `foreach` (not a big deal at all)

**Comment by andrewor14** [comment]
Ok merging into master 1.3 thanks @kayousterhout

**Pull request closed without merging**